DOCS-4036: Add more frame and geometry content, remove arm no code (#4411)
viamrobotics/docs: docs/operate/mobility/move-arm/frame-how-to.md

   You can think of it like this: The z-axis of the arm's origin frame is parallel to the z-axis of the world frame, and the origin of the arm's origin frame is in the same place as the origin of the world frame.

   If you want to define the world frame as the corner of the table, leave the default values for now, and continue to the next step so that you know which way `x` and `y` point and can edit the configuration accordingly.

1. Click **Save**.

   By configuring your arm's frame like this, you have now defined the world frame as the same as the arm's origin frame.

### Figure out which way the arm's axes point

The arm's origin frame is now the world frame, but you do not know which way the x, y, and z axes of the arm frame point, unless you happen to be familiar with the arm's [kinematics file](/operate/reference/kinematic-chain-config/).

1. On your arm's configuration card, click the **TEST** section (or use the **CONTROL** tab).

   {{% alert title="Caution" color="caution" %}}

> Moving the arm with the **TEST** or **CONTROL** panel sends commands using the arm API, which does not take obstacles or other components into account.
> Be careful not to hit people or objects with the arm.
> {{% /alert %}}

Author: ```suggestion
   Moving the arm with the **TEST** or **CONTROL** panel sends commands using the arm API, which does not take obstacles or other components into account.
   Be careful not to hit people or objects with the arm.
   {{% /alert %}}
```

---
title: "Configure your frame system"
linkTitle: "Configure your frame system"
weight: 20
type: "docs"
layout: "docs"
description: "Decide on and configure the frames of your arm, its workspace, and other components."
---

> Frames can be confusing.

Reviewer: I think we don't want to say that?
```suggestion
```

Author: #developerempathy? :D I will remove it

This guide will help you decide on a coordinate system for your workspace, and then determine how the frames of your arm and other components relate to that workspace.

Once you have configured frames for all your components, Viam will keep track of the positions and orientations as your robot moves, so you can plan motion in terms of a consistent coordinate system.

For reference information, see [the frame system](/operate/reference/services/frame-system/).

## Determine the world frame

The world reference frame is the fixed coordinate system that serves as the reference point for the other frames in your robotic system.

You choose the world frame, in whatever way is convenient for your application.
It generally makes sense to define the world frame's location as a point in your space that does not move and is easy to measure from.

> For example, if you are using a robot arm mounted to a table, it can be convenient to define the arm's base frame as equal to the world frame, or you can define the world frame as one corner of the table that the arm is mounted to.

Author: ```suggestion
For example, if you are using a robot arm mounted to a table, it can be convenient to define the arm's origin frame as equal to the world frame, or you can define the world frame as one corner of the table that the arm is mounted to.
```

> For example, if you are using a robot arm mounted to a table, it can be convenient to decide that the world frame origin is at the base of the arm, or you can define the world frame origin as one corner of the table that the arm is mounted to.

Author: ```suggestion
For example, if you are using a robot arm mounted to a table, it can be convenient to decide that the world frame origin is at the base of the arm, or you can decide that the world frame origin is one corner of the table that the arm is mounted to.
```

You do not explicitly configure the world frame.
You define it implicitly by configuring the frame of a component relative to world.

### Add a frame to your arm

1. Mount the arm in a fixed location.
1. On your arm's configuration card, click **+ Add frame**.
1. The card will populate with the following default values:

   ```json {class="line-numbers linkable-line-numbers"}
   {
     "orientation": {
       "type": "ov_degrees",
       "value": {
         "x": 0,
         "y": 0,
         "z": 1,
         "th": 0
       }
     },
     "parent": "world",
     "translation": {
       "x": 0,
       "y": 0,
       "z": 0
     }
   }
   ```

>    This [orientation vector](/operate/mobility/orientation-vector/) of `x=0, y=0, z=1, th=0` means the frame is aligned with the world frame, and the `0, 0, 0` translation means the arm's {{< glossary_tooltip term_id="origin-frame" text="origin frame" >}} is coincident with the origin of the world frame.

Reviewer: an image might help? optional

   You can think of it like this: The z-axis of the arm's origin frame is parallel to the z-axis of the world frame, and the origin of the arm's origin frame is in the same place as the origin of the world frame.

   If you want to define the world frame as the corner of the table, leave the default values for now, and continue to the next step so that you know which way `x` and `y` point and can edit the configuration accordingly.

1. Click **Save**.

   By configuring your arm's frame like this, you have now defined the world frame as the same as the arm's base frame.

### Figure out which way the arm's axes point

> The arm's base frame is now the world frame, but you do not know which way the x, y, and z axes of the arm frame point, unless you happen to be familiar with the arm's [kinematics file](/operate/reference/kinematic-chain-config/).

Reviewer 2: This selection of text references both the arm's "origin frame" and "base frame". These are the same thing, correct? Should we stick with just one of these names or explain the different if there is one? 

It's my understanding that that origin/base frame for an arm is references the first link (usually called a base link) in the kinematics chain, which typically aligns with the world. This is opposed to the end-effector frame when configuring a component like a gripper or camera frame in reference to a parent arm. I could be incorrect in my understanding though.

1. On your arm's configuration card, click the **TEST** section (or use the **CONTROL** tab).

> 

Author: ```suggestion

   {{% alert title="Note" color="note" %}}
Moving the arm with the **TEST** or **CONTROL** panel sends commands using the arm API, which does not take obstacles or other components into account.
Be careful not to hit people or objects with the arm.
{{% /alert %}}

```

1. Move the arm in each direction and note which way the arm moves.

>    For example, under the **MoveToPosition** section, move the arm in the positive Z direction.

Author: Added an alert!

   If it moves upwards, then you know that the z-axis of the arm frame points upwards.
1. If the directions do not match your intended world frame, you can edit the arm's `orientation` field to rotate the arm frame to match your intended world frame.
   For example, if the arm's x-axis points to the right, but you want it to point to the left, you can set the `orientation` field to `"x"=0, "y"=0, "z"=1, "th"=180`.

> 1. For your reference, mark the X, Y, and Z axes of the world frame on your workspace with tape or a marker.

Reviewer 2: It would be great to follow up with a picture of how to mark this as an example. Not a blocker.

> 1. For your reference, mark the X, Y, and Z axes of the world frame on your workspace with tape or a marker.

Reviewer: ```suggestion
1. For your reference, we recommend you mark the X, Y, and Z axes of the world frame on your workspace with tape or a marker.
```

### Add a frame offset

If you want to define the world frame as the corner of the table, you can now edit the arm's frame to be offset from the world frame by the distance from the table corner to the arm's base:

1. Measure the X and Y distance from your designated table corner to the center of the arm's base.
1. Edit the arm's frame configuration to set the `translation` field accordingly.
   For example, if the arm's base is 100mm in the X direction and 200mm in the Y direction from the table corner, set the `translation` field to `"x"=100, "y"=200, "z"=0`.
   If you chose not to align the arm's axes with the world axes, be sure to use the X, Y, and Z directions of the _world_ frame to set the `translation` field.
1. Click **Save**.

   The arm's frame is now offset from the world frame by the distance from the table corner to the arm's base.

> ## Other ways to define the world frame

Author: ```suggestion
## The world frame for mobile robots
```

> If none of your components are fixed in place, it might make sense to define the world frame as the frame of the first component you add to your robot, for example, the mobile base.

Reviewer 2: How is this different from how the arm's frame defines the world in the above example? Or is this something expected in code when using the motion service?

Author: @biotinker what would you suggest for wording here? Is this okay?

Reviewer 3: > The motion service will not automatically keep track of the relationship between the robot and its environment in this case.

Maybe mention the SLAM or Navigation service will be doing this instead, outside the framesystem configured in `app`?

Author: ```suggestion
If none of your components are fixed in place, the world frame is less meaningful.
	The motion service will not automatically keep track of the relationship between the robot and its environment in this case.
	The relative location of the robot will be tracked by the SLAM service or the navigation service, depending on which method you use.
	Neither of these services use the frame from your machine configuration.
	
	You can set the origin frame of one of your components (such as a {{< glossary_tooltip term_id="base" text="base" >}}) to `0, 0, 0` with respect to world, and parent other component frames to that component so that the motion service can prevent collisions between components.
```

> If none of your components are fixed in place, it might make sense to define the world frame as the frame of the first component you add to your robot, for example, the mobile base.

Reviewer: I think this should probably be earlier on this page when we talk about determining the world frame. Otherewise I have to get through a lot of info that may not suit me and I don't know that I need to get to this header. Maybe the header should be something about Mobile machines/robots ?

Author: Okay I'll change the header. However this is a less-well-supported case so I don't know that we want to highlight; see latest wording suggestion for why ^
